src/coreclr/vm/object.h

```diff
@@ -739,6 +739,78 @@ class PtrArray : public ArrayBase
 
 #define OFFSETOF__PtrArray__m_Array_              ARRAYBASE_SIZE
 
+/* Corresponds to the managed Span<T> and ReadOnlySpan<T> types.
+   This should only ever be passed from the managed to the unmanaged world byref,
+   as any copies of this struct made within the unmanaged world will not observe
+   potential GC relocations of the source data. As a result we've deleted ctors
+   which allow making byval copies of the span. */
+template < class KIND >
+class Span
+{
+private:
+    /* Keep fields below in sync with managed Span / ReadOnlySpan layout. */
+    KIND* _pointer;
+    unsigned int _length;
+
+public:
+    class ControlledMutabilityRef
+    {
+    private:
+        KIND* _pointer;
+
+    public:
+        __inline ControlledMutabilityRef(KIND* ptr)
+            : _pointer(ptr)
+        {
+        }
+
+        // Turns the pointer into a const ref, allowing safe dereferencing.
+        __inline operator const KIND&() const
+        {
+            return *_pointer;
+        }
+
+        // Forwards operator->() to the underlying element as non-const.
+        __inline KIND operator->() const
+        {
+            return *_pointer;
+        }
+    };
+
+    Span() = delete;
```

Comment:
Constructors change calling convention ABI. We try to keep these types that are passed on the FCall boundary as Plain Old Datatypes.

For example: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/qcall.h#L155

Reply:
I think this is desirable, though? I'm intentionally trying to force `Span<T>` taken byval in the ABI to cause a compilation error to avoid misuse, so the only legal way to pass it is byref or by pointer, neither of which involves ctor invocation. Contrast this with `QCall::StringHandleOnStack` and other types which are taken byval. Or do you think this is complicating things too much?

Reply:
It is fine to pass the Span as byval too, as long as you do not trigger GC or if you make sure that it is manually GC protected.